### ISSUE TITLE
Request: Sort iterables by

### ISSUE THREAD

**Issue by clintval** [erikrose/more-itertools#100]
I'm not sure if this is a fitting addition to `more-itertools` but it's a method I use quite often. This function sorts iterables using a defined order of priority. So you can sort iterables in concordance with a given sort pattern. I suppose it's tough to explain so here are three examples.

```python
# Will sort all iterables based on the ascending sort order of the first iterable
>>>sort_iterables_by([['a', 'd', 'c', 'd'], [1, 3, 2, 4]], key_list=(0,))
[('a', 'c', 'd', 'd'), (1, 2, 3, 4)]

# Will sort all iterables based on the ascending sort order of the first iterable,
# then the second iterable
>>>sort_iterables_by([['d', 'd', 'd', 'c'], [4, 3, 7, 10], [1, 2, 3, 4]],
                      key_list=(0, 1))
[('c', 'd', 'd', 'd'), (10, 3, 4, 7), (4, 2, 1, 3)]

# Will sort all iterables based on the descending sort order of the first iterable,
# then the second iterable
>>>sort_iterables_by([['a', 'b', 'b'], [1, 3, 2]],
>>>                   key_list=(0, 1),
>>>                   reverse=True))
[('b', 'b', 'a'), (3, 2, 1)]
```
Here is the function I propose
```python
import operator

def sort_iterables_by(iterables, key_list=(0,), reverse=False):

    return list(zip(*sorted(zip(*iterables),
                            key=operator.itemgetter(*key_list),
                            reverse=reverse)))
```
What do you guys think? A useful addition? One remark is that because `zip` is used, iterables are returned trimmed to the length of the shortest iterable before sorting. An alternate form of the function could  be used with `zip_longest` although for lists with heterogeneous objects no fillvalue will make obvious sense.

Example:

```python
import operator
import itertools

def sort_iterables_by(iterables, key_list=(0,), reverse=False, fillvalue=None):

    return list(zip(*sorted(itertools.zip_longest(*iterables, fillvalue=fillvalue),
                            key=operator.itemgetter(*key_list),
                            reverse=reverse)))
```

**Comment by bbayles** [comment]
I like it; I've done this myself before and have certainly seen the question posed on Stack Overflow. If you make a pull request with tests and a docstring I would be inclined to merge it.

Notes:
* I think there's probably a better name out there for this, maybe `zip_sort`?
* I'd like a test that has duplicates in one of the sort targets
* I'm good with trimming to the shortest iterable, as long as it's in the docstring.

**Comment by erikrose** [comment]
I like zip_sort except that result doesn't end up zipped. "Align" is another verb we could play with: align_sort? I'm not in love with it, but maybe it will somebody. Thanks, @clintval!

**Comment by clintval** [comment]
What do you all think about `priority_sort`? It's making sense to me since the function returns sorted objects in a priority that you define? `sort_iterables_by` is a bit wordy and not the clearest. `zip_sort` may not convey what the function is actually used for. I'll get on a PR regardless. Thanks for the input!

**Comment by erikrose** [comment]
Or `parallel_sort` or `sort_like`, to throw a few more into the pot. What's the "priority" in `priority_sort`?

**Comment by clintval** [comment]
You provide the parameter `key_list` which are the indexes of the iterables you wish to sort by (in priority order). So if you had 5 lists and you wanted them ordered using the sort pattern of the first list you would use `key_list=(0,)`. If you wanted them ordered using the sort pattern of the last list you would use `key_list=(4,)`.

If you had three lists and wanted them sorted based on the sort pattern of the first (affects all three), then the second (affects second and third), then the third (affects third only), you would use `key_list=(0, 1, 2)`.

Hope that clears it up!

This is most useful if you have a few lists containing classification information, etc.

|List 1|List 2|List 3|
|----|----|----|
|Georgia|Augusta|2000|
|Georgia|Atlanta|2100|
|Georgia|Macon|2100|
|Connecticut|Hartford|5500|
|Connecticut|Coventry|7200|
|Connecticut|Coventry|2300|

If you used `key_list=(0,)` you would get sorting based only on List 1:

|List 1|List 2|List 3|
|----|----|----|
|Connecticut|Hartford|5500|
|Connecticut|Coventry|7200|
|Connecticut|Coventry|2300|
|Georgia|Augusta|2000|
|Georgia|Atlanta|2100|
|Georgia|Macon|2100|

If you used `key_list=(0, 1)` you would get sorting based only on List 1, and then List 2:

|List 1|List 2|List 3|
|----|----|----|
|Connecticut|Coventry|7200|
|Connecticut|Coventry|2300|
|Connecticut|Hartford|5500|
|Georgia|Atlanta|2100|
|Georgia|Augusta|2000|
|Georgia|Macon|2100|

And finally, if you used `key_list=(0, 1, 2)` you would get sorting based only on List 1, then List 2, and then List 3:

|List 1|List 2|List 3|
|----|----|----|
|Connecticut|Coventry|2300|
|Connecticut|Coventry|7200|
|Connecticut|Hartford|5500|
|Georgia|Atlanta|2100|
|Georgia|Augusta|2000|
|Georgia|Macon|2100|

**Comment by bbayles** [comment]
Reading some Stack Overflow posts for ideas beased on how people talk about this problem:
* `argsort` is what numpy uses (but it works differently)
* `sort_together`
* `simul_sort` (or some diminutive for simultaneous)
* `multi_sort`

My favorite is `sort_together`.

**Comment by clintval** [comment]
Hi all,

I just submitted at #101. This is my first time contributing to an open source project (and the first time writing tests for anything other than my own use). Please do reach back out to me with advise if you have it! Thanks!

**Comment by bbayles** [comment]
Excellent, and congratulations! I will review the PR a little later.

**Comment by clintval** [comment]
Actually, bare with me. Did not run the doctests and they are not formatted correctly. I'll see what I can do to clean that one up!

**Comment by erikrose** [comment]
@bbayles In the SO posts you read, was there a compelling reason to go at the problem in situ like this rather than simply zipping the iterables together in the right order and then sorting? (Tuples sort lexically--that's the alternative I'm getting at.) The one that comes to my mind is that you want to maintain the original ordering of one of more of the iterables, which is doable since Python's sort() is stable. Otherwise, the purpose of the key_list arg can fulfilled just by passing the iterables in the right order.

**Comment by bbayles** [comment]
I think you're right about not wanting to mess with one of the iterables being legtimate. However, the most compelling case comes when there are un-orderable types in the iterables.

No key means this fails:
```python
>>> from operator import itemgetter
... A = 'ccb'
... B = [[None, None], [1, None], ['a', None]]
... 
... iterables = (A, B)
... list(zip(*sorted(zip(*iterables))))
Traceback (most recent call last):
  File "<pyshell#2>", line 6, in <module>
    list(zip(*sorted(zip(*iterables))))
TypeError: unorderable types: int() < NoneType()
```

With a key means it succeeds:
```pyhton
```

---

1. [How can I “zip sort” parallel numpy arrays?](http://stackoverflow.com/q/1903462/353839)
2. [python sorting two lists - zip, sort, zip - unorderable types: int() < NoneType()](http://stackoverflow.com/q/34621526/353839)
3. [Sorting a list in Python using the result from sorting another list](http://stackoverflow.com/q/9543211/353839)
4. [Sorting list based on values from another list?](http://stackoverflow.com/q/6618515/353839)
5. [Python sort one list from another list](http://stackoverflow.com/q/20514474/353839)
6. [python sorting based on parallel arrays](http://stackoverflow.com/q/17618243/353839)
7. [Python ordering the list according to another list](http://stackoverflow.com/q/18688820/353839)

(2) seems to be the most compelling case for including a "target keys" parameter.The user in (3) could also use it.

**Comment by clintval** [comment]
@erikrose I think the benefit to using the `key_list` param is that you can force the sort to consider only specific iterables. If you had data which looked like the table below, then you could sort based on List 1 without disturbing List 2's local ordering for the repeated elements in List 1 (one might want the color of cars to intentionally follow the scheme 'white', 'blue', 'green', but want the cars in alphabetical order).

|List 1|List 2|
|----|----|
|Honda|White|
|Honda|Blue|
|Honda|Green|
|Ford|White|
|Ford|Blue|
|Ford|Green|


After `sort_together([list1, list2], key_list=(0,))`:

|List 1|List 2|
|----|----|
|Ford|White|
|Ford|Blue|
|Ford|Green|
|Honda|White|
|Honda|Blue|
|Honda|Green|

For completeness you could sort both with `sort_together([list1, list2], key_list=(0, 1))` which, I imagine, would be the result of zipping and then sorting the zipped elements:

|List 1|List 2|
|----|----|
|Ford|Blue|
|Ford|Green|
|Ford|White|
|Honda|Blue|
|Honda|Green|
|Honda|White|

Since 'honda' comes before 'ford' in List 1 then this result is not surprising `sort_together([list1, list2], key_list=(1,))`:

|List 1|List 2|
|----|----|
|Honda|Blue|
|Ford|Blue|
|Honda|Green|
|Ford|Green|
|Honda|White|
|Ford|White|

Hope that helps to explain where I was coming from on this function. The above use case is similar to how I use it personally.

**Comment by erikrose** [comment]
Good enough for me! Thanks for the lengthy examples, @clintval.

**Comment by bbayles** [comment]
`sort_together()` was added in #103. Thanks @clintval! Do let us know your next idea.

**Comment by clintval** [comment]
Thanks all! Sorry for the late reply I've been away from my computer for the past few days. Excellent experience working with you on my first open source contribution. Thanks again.

**Comment by pylang** [comment]
Nice contrib @clintval.  Welcome.

**Comment by erikrose** [comment]
A pleasure to have you!
